lightweight-term-count-update.php

```diff
@@ -12,79 +12,219 @@
 		remove_action( 'transition_post_status', '_update_term_count_on_transition_post_status' );
-		add_action( 'transition_post_status', array( $this, 'quick_update_terms_count' ), 10, 3 );
-		$this->counted_stati = apply_filters( 'LTCU_counted_stati',  $this->counted_stati );
+
+		$this->counted_statuses = apply_filters( 'ltcu_counted_statuses', $this->counted_statuses );
+		add_action( 'transition_post_status', array( $this, 'transition_post_status' ), 10, 3 );
+		add_action( 'added_term_relationship', array( $this, 'added_term_relationship' ), 10, 3 );
+		add_action( 'deleted_term_relationships', array( $this, 'deleted_term_relationships' ), 10, 3 );
 	}
 
-	public function quick_update_terms_count( $new, $old, $post ) {
-		global $wpdb;
+	/**
+	 * When a term relationship is added, increment the term count.
+	 *
+	 * @param int    $object_id Object ID.
+	 * @param int    $tt_id     Single term taxonomy ID.
+	 * @param string $taxonomy  Taxonomy slug.
+	 */
+	public function added_term_relationship( $object_id, $tt_id, $taxonomy ) {
+		$post = get_post( $object_id );
+		if ( in_array( $post->post_status, $this->counted_statuses, true ) ) {
+			$this->quick_update_terms_count( $post, (array) $tt_id, $taxonomy, 'increment' );
+		}
+	}
+
+	/**
+	 * When a term relationship is deleted, decrement the term count.
+	 *
+	 * @param int    $object_id Object ID.
+	 * @param array  $tt_ids    Array of term taxonomy IDs.
+	 * @param string $taxonomy  Taxonomy slug.
+	 */
+	public function deleted_term_relationships( $object_id, $tt_ids, $taxonomy ) {
+		$post = get_post( $object_id );
+		$this->quick_update_terms_count( $post, $tt_ids, $taxonomy, 'decrement' );
+	}
+
+	/**
+	 * When a post transitions, increment or decrement term counts as
+	 * appropriate.
+	 *
+	 * @param  string   $new_status New post status.
+	 * @param  string   $old_status Old post status.
+	 * @param  \WP_Post $post       Post being transitioned.
+	 */
+	public function transition_post_status( $new_status, $old_status, $post ) {
+		foreach ( (array) get_object_taxonomies( $post->post_type ) as $taxonomy ) {
+			$tt_ids = wp_get_object_terms( $post->ID, $taxonomy, array(
+				'fields' => 'tt_ids',
+			) );
+
+			if ( ! empty( $tt_ids ) ) {
+				$this->quick_update_terms_count(
+					$post,
+					$tt_ids,
+					$taxonomy,
+					$this->transition_type( $new_status, $old_status )
+				);
+			}
+		}
+
+		// For non-attachments, let's check if there are any attachment children
+		// with inherited post status -- if so those will need to be re-counted.
+		if ( 'attachment' !== $post->post_type ) {
+			$attachments = new WP_Query( array(
+				'post_type' => 'attachment',
+				'post_parent' => $post->ID,
+				'post_status' => 'inherit',
+			) );
 
-		$transition_type = $this->transition_type( $new, $old );
+			if ( $attachments->have_posts() ) {
+				foreach ( $attachments->posts as $post ) {
+					$this->transition_post_status( $new_status, $old_status, $post );
```

Comment:
I was trying to cover this situation in https://github.com/Automattic/lightweight-term-count-update/pull/2/files#diff-753396fe882316d54c2c4341cc3ab424R231 which is currently failing. Let me dig into why is that.

NB: I have tested all the unit tests on a standard WordPress install w/o the plugin being used and they all are passing.